File: src/main/java/editor/filter/leaf/options/single/BlockFilter.java
package editor.filter.leaf.options.single;

import java.util.HashSet;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import editor.database.attributes.CardAttribute;
import editor.filter.Filter;

/**
 * This class represents a filter that groups cards by block.
 *
 * @author Alec Roelke
 */
public class BlockFilter extends SingletonOptionsFilter<String>
{
    /**
     * Create a new BlockFilter.
     */
    public BlockFilter()
    {
        super(CardAttribute.BLOCK, (c) -> c.expansion().block);
    }

    @Override
    protected String convertFromString(String str)
    {
        return str;
    }

    @Override
    public Filter copy()
    {
        BlockFilter filter = (BlockFilter)CardAttribute.createFilter(CardAttribute.BLOCK);
        filter.contain = contain;
        filter.selected = new HashSet<>(selected);
        return filter;
    }

    @Override
    protected JsonElement convertToJson(String item)
    {
        return new JsonPrimitive(item);
    }

    @Override
    protected String convertFromJson(JsonElement item)
    {
        return item.getAsString();
    }
}

File: src/main/scala/editor/filter/leaf/options/single/BlockFilter.scala
package editor.filter.leaf.options.single

import editor.database.attributes.CardAttribute
import editor.filter.FaceSearchOptions
import editor.util.Containment

/**
 * A filter that groups cards by the block they belong in.
 * @author Alec Roelke
 */
final case class BlockFilter(contain: Containment = Containment.AnyOf, selected: Set[String] = Set.empty) extends SingletonOptionsFilter[String, BlockFilter] {
  override def faces = FaceSearchOptions.ANY
  override def attribute = CardAttribute.Block
  override val unified = true
  override def value = _.expansion.block
  override def copy(faces: FaceSearchOptions, contain: Containment, selected: Set[String]) = copy(contain = contain, selected = selected)
}